keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+ 
+```
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                                // because we lack {,} , this is a single Peer selecting
+                                // pods, from namespaces selected by namespace selector.
+                                // This is difficult to verify for correctness at a
+                                // glance, due to the verbosity of the struct.
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+```
+
+In contrast, we can express the same and test using the API (including its entire connectivity matrix) in this proposal as follows:
+
+```
+	builder := &NetworkPolicySpecBuilder{}
+	builder = builder.SetName("myns", "allow-podb-in-nsb").SetPodSelector(map[string]string{"pod": "b"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod-name": "b"}, nil, map[string]string{"ns-name": "b"}, nil)
+	policy := builder.Get()
+	reachability := NewReachability(allPods, false)
+	reachability.ExpectAllIngress(Pod("myns/b"), false)
+	reachability.Expect(Pod("b/b"), Pod("myns/b"), true)
+	reachability.Expect(Pod("myns/b"), Pod("myns/b"), true)
+```
+We can of course make this much easier to reuse and reason about, as well as make it self-documenting, and will outline how in the solutions section.
+
+### Performance
+ 
+CURRENTLY: For each test, a few pods are spun up, and a polling process occurs where we wait for the pod to complete successfully and report it could connect to a target pod.  Because all clusters start pods at different rates, heuristics have to be relied on for timing a test out.  A large, slow cluster may not be capable of spinning pods up quickly, and thus may timeout one of the 25 tests, leading to a false negative result.
+ 
+In some clusters, for example, namespace deletion is known to be slow - and in these cases the network policy tests may take more than an hour to complete.
+ 
+- If network policies or pod CIDR's are not correct, it's likely all tests can fail, and thus the network policy suite may take an hour to finish, based on the estimate of 3 minutes, for each failed test, alongside 25 tests (in general, NetworkPolicy tests on a healthy EC2 cluster, with no traffic and broken network policies, take between 150 and 200 seconds complete).
```

Comment:
"It takes a long time to fail if we accidentally break the tests" is not a strong argument

Reply:
fair point, will keep this bullet for now though since it's still i think relevant to some users